content/2017-03-15-criando-um-aplicativo-de-linha-de-comando-com-python.markdown

```diff
@@ -0,0 +1,539 @@
+    args.second_number,
+    args.first_number + args.second_number
+))
+```
+O exemplo de uso do código anterior seria algo como:
+```
+$ python cli.py 2 5
+2 + 5 = 7
+```
+> Não se preocupe com o `type` no exemplo, veremos o que isso significa mais adiante
+
+#### Argumentos opcionais
+Os argumentos opcionais são declarados com um ou dois hífens no prefixo do nome (e.g. _-f_, _--foo_) e não
+dependem de uma posição especifica para serem informados.
+Usando o exemplo anterior, vamos adicionar um argumento opcional para determinar se o output do comando
+deve ou não ser verboso:
+
+```python
+# cli.py
+from argparse import ArgumentParser
+
+
+parser = ArgumentParser()
+parser.add_argument('first_number', type=int)
+parser.add_argument('second_number', type=int)
+parser.add_argument('--verbose', action='store_true')
+
+args = parser.parse_args()
+result = args.first_number + args.second_number
+if args.verbose:
+    print('{} + {} = {}'.format(
+        args.first_number,
+        args.second_number,
+        result
+    ))
+else:
+    print(result)
+```
+Desta forma, possibilitamos saidas diferentes da nossa aplicação de acordo com a presença ou não do
+argumento `--verbose`.
+
+```
+$ python cli.py 2 5
+7
+$ python t.py 2 5 --verbose
+2 + 5 = 7
+```
+> Não se preocupe com o `action` no exemplo, veremos o que isso significa mais adiante
+
+Podemos simplificar aindas mais o argumento _--verbose_ criando uma opção encurtada dele.
+O método `add_argument` do `ArgumentParser` aceita uma lista de nomes do argumento, sendo assim,
+basta adicionar as opções na criação do argumento:
+
+```python
+parser.add_argument('-v', '--verbose', action='store_true')
+```
+E inclusive, essa alteração já reflete no _help_ da aplicação:
+
+```
+$ python cli.py -h
+usage: cli.py [-h] [-v] first_number second_number
+
+positional arguments:
+  first_number
+  second_number
+
+optional arguments:
+  -h, --help     show this help message and exit
+  -v, --verbose
+```
+
+Vale ressaltar que essa opção só faz sentido para argumentos opcionais.
+
+### Types
+Por padrão, o valor todos os argumentos são interpretados como _string_, porém, é possível determinar um tipo
+para esses valores, através do parametro `type` do método `add_argument` como já vimos anteriormente.<br>
+Ao determinar um _type_ para um argumento, será executado um _type-checking_ no momento do *parse_args()* para garantir
+que o valor do argumento é do tipo especificado:
+
+````python
+>>> from argparse import ArgumentParser
+>>> parser = ArgumentParser()
+>>> parser.add_argument('number', type=int)
+>>> args = parser.parse_args(['a'])
+usage: [-h] number
+: error: argument number: invalid int value: 'a'
+```
+Qualquer _callable_ pode ser usado como type, pontencializando ainda mais o uso de tipos nos argumentos, e.g.
+
+```python
+# cli.py
+import argparse
+
+
+def odd(n):
+    n = int(n)
+    if n % 2 == 1:
+        return n
+    raise argparse.ArgumentTypeError('{} its not an odd number'.format(n))
```

Comment:
Aqui acho que só `it's not` ao invés de `its not` (que se não me engano, `its` é pronome possessivo).